lib/widgets/editor/editor_toolbar.dart

```diff
@@ -0,0 +1,462 @@
+              ),
+          ],
+          onSelected: (val) {
+            final header = '${'#' * val.value} ';
+
+            if (!controller.firstSelectedLine.startsWith(header)) {
+              controller.insertAtBeginningOfFirstSelectedLine(header);
+            }
+          },
+          tooltip: L10n.of(context).editor_header,
+          child: const Icon(Icons.h_mobiledata),
+        ),
+        IconButton(
+          onPressed: () => controller.surround(
+            before: '~~',
+            placeholder: L10n.of(context).insert_text_here_placeholder,
+          ),
+          icon: const Icon(Icons.format_strikethrough),
+          tooltip: L10n.of(context).editor_strikethrough,
+        ),
+        IconButton(
+          onPressed: () {
+            controller.insertAtBeginningOfEverySelectedLine('> ');
+          },
+          icon: const Icon(Icons.format_quote),
+          tooltip: L10n.of(context).editor_quote,
+        ),
+        IconButton(
+          onPressed: () {
+            final line = controller.firstSelectedLine;
+
+            if (line.startsWith(RegExp.escape('* '))) {
+              controller.removeAtBeginningOfEverySelectedLine('* ');
+            } else if (line.startsWith('- ')) {
+              controller.removeAtBeginningOfEverySelectedLine('- ');
+            } else {
+              controller.insertAtBeginningOfEverySelectedLine('- ');
+            }
+          },
+          icon: const Icon(Icons.format_list_bulleted),
+          tooltip: L10n.of(context).editor_list,
+        ),
+        IconButton(
+          onPressed: () => controller.surround(
+            before: '`',
+            placeholder: L10n.of(context).insert_text_here_placeholder,
+          ),
+          icon: const Icon(Icons.code),
+          tooltip: L10n.of(context).editor_code,
+        ),
+        IconButton(
+          onPressed: () => controller.surround(
+            before: '~',
+            placeholder: L10n.of(context).insert_text_here_placeholder,
+          ),
+          icon: const Icon(Icons.subscript),
+          tooltip: L10n.of(context).editor_subscript,
+        ),
+        IconButton(
+          onPressed: () => controller.surround(
+            before: '^',
+            placeholder: L10n.of(context).insert_text_here_placeholder,
+          ),
+          icon: const Icon(Icons.superscript),
+          tooltip: L10n.of(context).editor_superscript,
+        ),
+        //spoiler
+        IconButton(
+          onPressed: () {
+            controller.reformat((selection) {
+              final insides = selection.isNotEmpty ? selection : '___';
+              Logger.root
+                  .info([21, 21 + insides.length, insides, insides.length]);
+              return Reformat(
+                text: '\n::: spoiler spoiler\n$insides\n:::\n',
+                selectionBeginningShift: 21,
+                selectionEndingShift: 21 + insides.length - selection.length,
```

Comment:
What is this magical number?

Reply:
length of the string above '\n::: spoiler spoiler\n...\n:::\n'

Reply:
Then it should be computed off that string